[RFC] Add sparsehash support for hashed dictionary (to reduce memory usage)

azat opened the pull request:
- [Benchmarks of various hash tables](https://github.com/yandex/ClickHouse/pull/6894#issuecomment-532570298)

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):

Introduce `<sparse_hashed>` layout for dictionaries, that will make dictionary uses less memory.

Detailed description (optional):

This patch introduces separate attribute for dictionaries -- <sparse>

With sparse enabled, sparsehash (actually sparsehash-c11, that do not
requires sparseconfig.h, for easier packaging with cmake) will be used
over default HashMap, sparsehash is more memory efficient but it is also
slower.

So in a nutshell:
- HashMap uses ~2x more memory then sparse_hash_map
- HashMap ~2-2.5x faster then sparse_hash_map
(tested on lots of input, and the most close to production was
dictionary with 600KK hashes and UInt16 as value)

alexey-milovidov commented:
The idea is very reasonable, I also had some thoughts about it.

You can also check [F14](https://github.com/facebook/folly/blob/master/folly/container/F14.md) as a memory efficient hash table.

alexey-milovidov commented:
As we already use sparsehash/densehash
https://github.com/yandex/ClickHouse/tree/master/contrib/libsparsehash
maybe it's better to unify it and use single library (remove old library)?

alexey-milovidov commented:
It's in question - should we provide the choice of the data structure via attribute or via different `layout`. I think, different layout is better.

Another point: current data structure is larger than needed, because it uses separate hash tables for each dictionary attribute. It's possible to use single data structure similar to what we have in `Join.h`.

Using the same data structure for Join is a great way to unify support for all data types (but it will also lead to worse performance).

data-man commented:
Compare with [Tessil](https://github.com/Tessil)'s collections. ([results](https://tessil.github.io/2016/08/29/benchmark-hopscotch-map.html))
Especially with [hat-trie](https://github.com/Tessil/hat-trie) for strings. ([results](https://github.com/Tessil/hat-trie#benchmark))

azat commented:
> You can also check F14 as a memory efficient hash table.

I will take a look, but seems that sparsehash is more memory efficient anyway.

>maybe it's better to unify it and use single library (remove old library)?

Sure, will take a look

>It's in question - should we provide the choice of the data structure via attribute or via different layout. I think, different layout is better.

Agree, since in case of attributes it simply ignores non existing, so you cannot be sure what hashtable will be used, I will update the patch

> Another point: current data structure is larger than needed, because it uses separate hash tables for each dictionary attribute.

That my second concern, but this can/should be done separately anyway (objections?)

> It's possible to use single data structure similar to what we have in Join.h.
> Using the same data structure for Join is a great way to unify support for all data types (but it will also lead to worse performance).

Hm I don't like this part
>but it will also lead to worse performance

> Compare with Tessil's collections.

Will take a look, thanks!

alexey-milovidov commented:
> That my second concern, but this can/should be done separately anyway (objections?)

Ok.

akuzm commented:
Oops, I missed this.

So, the idea is that our HashMap uses too much memory because of its exponential resize, so we use a sparse hash map that doesn't allocate as much empty memory? 
Does our TwoLevelHashMap also have better memory usage?

It would be good to have a reference dictionary and queries, so that we can measure the memory usage and the performance, and compare different implementations.

akuzm commented:
On a second thought, the memory usage of two-level table should only be different when the hash is not uniform and the occupancy of buckets is significantly different. For an ideal hash, memory usage should be the same.

alexey-milovidov commented:
@akuzm Our hash table grows 2x and have max fill factor of 50%. The memory overhead is 2 to 4 times.

Sparsehash memory overhead is very close to nothing and it's clearly better.

Other options is to use robin-hood probing that allows to keep max fill factor about 90%*. But it will still have 2x overhead. Also look at the Facebook's F14.

(Please note that it cannot beat our HashTable in our scenario even with 50% fill factor)

azat commented:
So here are benchmark results (for `uint64_t` key/value pair and 1e7 keys):

| container                 | insert   | find    | maxrss |
|---------------------------|----------|---------|--------
| `google::sparse_hash_map` | 8.99111  | 1.77821 | 212MiB |
| `tsl::sparse_map`         | 4.03314  | 1.67305 | 214MiB |
| `folly::F14VectorMap`     | 3.03167  | 1.19759 | 265MiB |
| `folly::F14FastMap`       | 2.52551  | 1.28362 | 393MiB |
| `folly::F14NodeMap`       | 3.42903  | 1.23989 | 441MiB |
| `google::dense_hash_map`  | 1.80534  | 1.17399 | 771MiB |

Looks like the most promising is `tsl::sparse_map`, thoughts?
Maybe google::sparse_hash_map for now (since it is well known and already used by ch), and switch to tsl::sparse_map later?

alexey-milovidov commented:
> Maybe google::sparse_hash_map for now (since it is well known and already used by ch), and switch to tsl::sparse_map later?

Ok.

PS. How did you test insert and find speed? What is the benchmark?

azat commented:
>PS. How did you test insert and find speed? What is the benchmark?

In a nutshell they are looks like:

```c
static uint64_t constexpr KEYS = 1e7;

template <class C>
void insert(C &c)
{
    for (uint64_t i = 0; i < KEYS; ++i) {
        c.insert(std::make_pair(crc64(i), i));
    }
}

template <class C>
void find(C &c)
{
    uint64_t found = 0;
    for (uint64_t i = 0; i < KEYS; ++i) {
        found += (c.find(crc64(i)) != c.end());
    }
    if (found != KEYS) {
        throw std::runtime_error("mismatch");
    }
}
```

Those `insert`/`find` was called for:
- `google::sparse_hash_map<uint64_t, uint64_t>`
- `folly::F14VectorMap<uint64_t, uint64_t>`
- ...

If you are interested in the whole benchmark, let me know (I did not share them yet since I need to fix some absolute includes there).

silviucpp commented:
@azat  have you tested also https://github.com/martinus/robin-hood-hashing it claims to perform better than more other : https://martin.ankerl.com/2019/04/01/hashmap-benchmarks-01-overview/

azat commented:
[Added `robin_hood::unordered_map`](https://github.com/yandex/ClickHouse/pull/6894#issuecomment-532570298)

data-man commented:
@azat I forgot about [parallel-hashmap](https://github.com/greg7mdp/parallel-hashmap) :smile:
Based on [abseil](https://github.com/abseil/abseil-cpp)'s hashes, but absl is a big library.

azat commented:
>I forgot about parallel-hashmap

[Added](https://github.com/yandex/ClickHouse/pull/6894#issuecomment-532570298)

P.S. parallel-hashmap is not about memory efficiency, so I added sparsepp too

azat commented:
> >It's in question - should we provide the choice of the data structure via attribute or via different layout. I think, different layout is better.

> Agree, since in case of attributes it simply ignores non existing, so you cannot be sure what hashtable will be used, I will update the patch

@alexey-milovidov resolved (as well as coding style issues)

alexey-milovidov commented:
@azat The test is decent. BTW, what hash function is used in sparsehash by default? If it's `std::hash` (trivial hash function for integers and rather inferior hash function for strings), you should not forget to provide our hash function as we do with HashMap.

azat commented:
>what hash function is used in sparsehash by default? If it's std::hash 

Indeed:

```
$ gg std::hash
sparsehash/dense_hash_map:template <class Key, class T, class HashFcn = std::hash<Key>,
sparsehash/dense_hash_set:template <class Value, class HashFcn = std::hash<Value>,
sparsehash/sparse_hash_map:template <class Key, class T, class HashFcn = std::hash<Key>,
sparsehash/sparse_hash_set:template <class Value, class HashFcn = std::hash<Value>,
```

>you should not forget to provide our hash function

Fixed

azat commented:
(fixed compilation error, due to missing `target_include_directories`)

azat committed:
Add new dictionary layout (sparse_hashed) that is more memory efficient

With this new layout, sparsehash will be used over default HashMap,
sparsehash is more memory efficient but it is also slower.

So in a nutshell:
- HashMap uses ~2x more memory then sparse_hash_map
- HashMap ~2-2.5x faster then sparse_hash_map
(tested on lots of input, and the most close to production was
dictionary with 600KK hashes and UInt16 as value)

TODO:
- fix allocated memory calculation
- getBufferSizeInBytes/getBufferSizeInCells interface
- benchmarks

v0: replace HashMap with google::sparse_hash_map
v2: use google::sparse_hash_map only when <sparse> isset to true
v3: replace attributes with different layout
v4: use ch hash over std::hash

akuzm commented on docs/en/query_language/dicts/external_dicts_dict_layout.md:
````diff
@@ -77,6 +78,18 @@ Configuration example:
 </layout>
 ```
 
+### sparse_hashed {#dicts-external_dicts_dict_layout-sparse_hashed}
````
So we're adding the usage of sparse ht as a dictionary-wide setting, not attribute-wise? This makes sense, if we're going to change it to the one hash table per dictionary later. 
This makes me wonder -- do we really need `if (sparse)` everywhere? Instead of this, I'd add a hash table template parameter to the HashedDictionary -- no runtime conditions and much less changes in code.

azat replied:
>Instead of this, I'd add a hash table template parameter to the HashedDictionary -- no runtime conditions and much less changes in code.

This was my first attempt, but then I decided that there will be too much templating

alexey-milovidov replied:
`if (sparse)` is dirty but tolerable.

akuzm commented on dbms/src/Dictionaries/HashedDictionary.cpp:
```diff
@@ -3,6 +3,23 @@
+{
+
+/// google::sparse_hash_map
+template <class T> auto first(const T &lhs) -> decltype(lhs.first)
```
Wow, this is my first time seeing the trailing return type syntax. You're not converting the return value to a different type, so it doesn't add anything over just `auto`, and can be omitted.

azat replied:
Nope, this is SFINAE, well not exactly, but the same purpose.

`HashMap` provides:
- `getFirst()`
- `getSecond()`

`google::sparse_hash_map` provides:
- `.first`
- `.second`

So we need to use  `getFirst()`/`getSecond()` only if we have it otherwise `.first`/`.second`

alexey-milovidov replied:
Ok. BTW, we cannot tolerate such tricks without a clarifying comment in code (added).

akuzm commented on dbms/src/Dictionaries/HashedDictionary.h:
```diff
@@ -186,6 +193,23 @@ class HashedDictionary final : public IDictionary
             CollectionPtrType<Float64>,
             CollectionPtrType<StringRef>>
             maps;
+        std::variant<
```
You probably won't have to change this after adding the collection type parameter, but if you do, consider using a double-macro approach (it has a name I forgot), like `APPLY_FOR_AGGREGATED_VARIANTS` in Aggregator.cpp.

azat replied:
>APPLY_FOR_AGGREGATED_VARIANTS

Like this?

```c
#define APPLY_FOR_TYPE(t)
    CollectionPtrType<t>,\
    SparseCollectionPtrType<t>\

  APPLY_FOR_TYPE(UInt8),
  ...
  APPLY_FOR_TYPE(StringRef)
#undef APPLY_FOR_TYPE
```

I was thinking about this, but decided that separate sparse_maps is cleaner, but I don't have strong filing against this.

So the question is do you want `sparse` to be class template argument or not?

akuzm reviewed (changes requested):
Left some inline comments about what could be improved.

Also, could you publish the benchmark so that I can reproduce your results? Maybe it even makes sense to save it for the future in Dictionary/tests, like we do with synthetic hash table benchmarks in Interpreters/tests.

azat commented:
> could you publish the benchmark so that I can reproduce your results?

For now it is here:

  https://github.com/azat-archive/hashtable-bench.git

  (although it requires some #ifdef cleanups)

> Maybe it even makes sense to save it for the future in Dictionary/tests

Hm, this will requires even more submodules, but unless you are fine
with this I'm fine too.
Or maybe just a bench HashMap vs google::sparse_hash_map?

alexey-milovidov committed:
Update HashedDictionary.cpp

alexey-milovidov committed:
Comments are the must.

alexey-milovidov committed:
Update HashedDictionary.h

alexey-milovidov commented:
The only way to provide proper performance tests (integrated in CI) is to finish https://github.com/yandex/ClickHouse/pull/4178 first.

azat committed:
Update HashedDictionary.cpp (fixed rename in return type for SFINAE)

azat commented:
@alexey-milovidov objections on squashing the following commits?

```
a12cffee91 Update HashedDictionary.h
ab9a7be45a Comments are the must.
9afea6f0b3 Update HashedDictionary.cpp
```

alexey-milovidov commented:
Yes, you can always squash my commits in your PR.
I merged this PR because I don't care too much about the number of commits.

azat commented:
> I don't care too much about the number of commits.

I don't care about number of commits either, I'm care about history, but until you are fine with this, I'm okay.

>merged

Thanks!
